traits/trait_dict_object.py

```diff
@@ -9,46 +9,39 @@
 # Thanks for using Enthought open source!
 
 import copy
-import logging
 from weakref import ref
 
-from .trait_base import Undefined
-from .trait_errors import TraitError
-
-# Set up a logger:
-logger = logging.getLogger(__name__)
+from traits.trait_base import Undefined
+from traits.trait_errors import TraitError
 
 
 class TraitDictEvent(object):
     """ An object reporting in-place changes to a traits dict.
 
     Parameters
     ----------
-    added : dict or None
-        New keys and values, or optionally None if nothing was added.
-    changed : dict or None
-        Updated keys and their previous values, or optionally None if nothing
-        was changed.
-    removed : dict or None
-        Old keys and values that were just removed, or optionally None if
-        nothing was removed.
+    removed : dict, optional
+        Old keys and values that were just removed.
+    added : dict, optional
+        New keys and values that were just added.
+    changed : dict, optional
+        Updated keys and their previous values.
 
     Attributes
     ----------
+    removed : dict
+        Old keys and values that were just removed.
     added : dict
-        New keys and values.  If nothing was added this is an empty dict.
+        New keys and values that were just added.
     changed : dict
-        Updated keys and their previous values.  If nothing was changed this
-        is an empty dict.
-    removed : dict
-        Old keys and values that were just removed.  If nothing was removed
-        this is an empty dict.
+        Updated keys and their previous values.
     """
 
-    def __init__(self, added=None, changed=None, removed=None):
-        # Construct new empty dicts every time instead of using a default value
-        # in the method argument, just in case someone gets the bright idea of
-        # modifying the dict they get in-place.
+    def __init__(self, removed=None, added=None, changed=None):
```

Comment:
This is potentially a backwards-incompatible change: if there's any code out there that creates an event with `TraitDictEvent(added, changed, removed)`, it'll have its meaning subtly changed. I can't _find_ any such code, but that doesn't mean that it doesn't exist.

I see two options: (1) restore the signature to its original, introducing a mismatch with the notification signature, or (2) make `added`, `changed` and `removed` keyword-only arguments, which at least ensures that anyone mis-using `TraitDictEvent` gets a hard error instead of a subtly-changed meaning (which could result in hard-to-diagnose errors). (All uses I can find outside this module already pass by keyword here.)